kad-dht/README.md

```diff
@@ -0,0 +1,453 @@
+# libp2p Kademlia DHT specification
+
+| Lifecycle Stage | Maturity       | Status | Latest Revision |
+|-----------------|----------------|--------|-----------------|
+| 3A              | Recommendation | Active | r0, 2021-05-07  |
+
+Authors: [@raulk], [@jhiesey], [@mxinden]
+
+Interest Group:
+
+[@raulk]: https://github.com/raulk
+[@jhiesey]: https://github.com/jhiesey
+[@mxinden]: https://github.com/mxinden
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+---
+
+## Overview
+
+The Kademlia Distributed Hash Table (DHT) subsystem in libp2p is a DHT
+implementation largely based on the Kademlia [0] whitepaper, augmented with
+notions from S/Kademlia [1], Coral [2] and the [BitTorrent DHT][bittorrent].
+
+This specification assumes the reader has prior knowledge of those systems. So
+rather than explaining DHT mechanics from scratch, we focus on differential
+areas:
+
+1. Specialisations and peculiarities of the libp2p implementation.
+2. Actual wire messages.
+3. Other algorithmic or non-standard behaviours worth pointing out.
+
+For everything else that isn't explicitly stated herein, it is safe to assume
+behaviour similar to Kademlia-based libraries.
+
+Code snippets use a Go-like syntax.
+
+## Definitions
+
+### Replication parameter (`k`)
+
+The amount of replication is governed by the replication parameter `k`. The
+default value for `k` is 20.
+
+### Distance
+
+In all cases, the distance between two keys is `XOR(sha256(key1),
+sha256(key2))`.
+
+### Kademlia routing table
+
+An implementation of this specification must try to maintain `k` peers with
+shared key prefix of length `L`, for every `L` in `[0..(keyspace-length - 1)]`,
+in its routing table. Given the keyspace length of 256 through the sha256 hash
+function, `L` can take values between 0 (inclusive) and 255 (inclusive). The
+local node shares a prefix length of 256 with its own key only.
+
+Implementations may use any data structure to maintain their routing table.
+Examples are the k-bucket data structure outlined in the Kademlia paper [0] or
+XOR-tries (see [go-libp2p-xor]).
+
+### Alpha concurrency parameter (`α`)
+
+The concurrency of node and value lookups are limited by parameter `α`, with a
+default value of 3. This implies that each lookup process can perform no more
+than 3 inflight requests, at any given time.
+
+## DHT operations
+
+The libp2p Kademlia DHT offers the following types of operations:
+
+- **Peer routing**
+
+  - Finding the closest nodes to a given key via `FIND_NODE`.
```

Comment:
> On FIND_NODE a DHT server returns a single DHT client if and only if the DHT client's peer ID matches the target.

Not quite, if we happen to know about a peer whose peerID is an exact match for the FIND_NODE query key then we add them to our list of servers to return. https://github.com/libp2p/go-libp2p-kad-dht/blob/6fff2a3b391f73da7a1c7558e27725ebe730e7ba/handlers.go#L256

> In my eyes documenting this distinction without documenting the DHT client / server mode extension (see your comment on rust-libp2p) would be inconsistent.

I guess that's fair. If all nodes were servers this distinction wouldn't really matter. A follow up PR is fine. Note: client mode isn't just for unreachable peers, it's generally recommended for "low quality" peers. This could mean low power, low bandwidth, low network uptime, etc. Basically peers that if they were servers would do more harm then good probably shouldn't be servers but may still want to make queries.